lensless/unrolled_admm.py

```diff
@@ -0,0 +1,254 @@
+        mu1=1e-6,
+        mu2=1e-5,
+        mu3=4e-5,
+        tau=0.0001,
+        psi=None,
+        psi_adj=None,
+        psi_gram=None,
+        **kwargs
+    ):
+        """
+
+        Parameters
+        ----------
+        psf : :py:class:`~torch.Tensor` of shape (H, W, C)
+            Point spread function (PSF) that models forward propagation.
+        dtype : float32 or float64
+            Data type to use for optimization. Default is float32.
+        n_iter : int, optional
+            Number of iterations to unrolled, by default 5
+        mu1 : float
+            Initial step size for updating primal/dual variables.
+        mu2 : float
+            Initial step size for updating primal/dual variables.
+        mu3 : float
+            Initial step size for updating primal/dual variables.
+        tau : float
+            Initial weight for L1 norm of `psi` applied to the image estimate.
+        psi : :py:class:`function`, optional
+            Operator to map image to a space that the image is assumed to be
+            sparse in (hence L1 norm). Default is to use total variation (TV)
+            operator.
+        psi_adj : :py:class:`function`
+            Adjoint of `psi`.
+        psi_gram : :py:class:`function`
+            Function to compute gram of `psi`.
+        """
+
+        # call reset() to initialize matrices
+
+        super(unrolled_ADMM, self).__init__(
+            psf, n_iter=n_iter, dtype=dtype, pad=False, norm="backward"
+        )
+
+        self._mu1 = torch.nn.Parameter(torch.ones(self.n_iter, device=self._psf.device) * mu1)
+        self._mu2 = torch.nn.Parameter(torch.ones(self.n_iter, device=self._psf.device) * mu2)
+        self._mu3 = torch.nn.Parameter(torch.ones(self.n_iter, device=self._psf.device) * mu3)
+        self._tau = torch.nn.Parameter(torch.ones(self.n_iter, device=self._psf.device) * tau)
+
+        # set prior
+        if psi is None:
+            # use already defined Psi and PsiT
+            self._PsiTPsi = finite_diff_gram(self._padded_shape, self._dtype, self.is_torch)
+        else:
+            assert psi_adj is not None
+            assert psi_gram is not None
+            assert callable(psi)
+            assert callable(psi_adj)
+            assert callable(psi_gram)
+            # overwrite already defined Psi and PsiT
+            self._Psi = psi
+            self._PsiT = psi_adj
+            self._PsiTPsi = psi_gram(self._padded_shape)
+
+        self._PsiTPsi = self._PsiTPsi.to(self._psf.device)
+
+        self.reset()
+
+    def _Psi(self, x):
+        """
+        Operator to map image to space that the image is assumed to be sparse
+        in.
+        """
+        return finite_diff(x)
+
+    def _PsiT(self, U):
+        """
+        Adjoint of `_Psi`.
+        """
+        return finite_diff_adj(U)
+
+    def reset(self, batch_size=1):
+        # needed because ReconstructionAlgorithm initializer call reset to early
+        if not hasattr(self, "_mu1"):
```

Comment:
Should we remove the call to reset from ReconstructionAlgorithm constructor?
Each algorithm would the be responsible to call reset if and when needed.